presto-client/src/main/java/io/prestosql/client/ErrorLocation.java

```diff
@@ -26,17 +27,20 @@
 {
     private final int lineNumber;
     private final int columnNumber;
+    private final String semanticErrorCodeString;
```

Comment:
Why not to add this to `ExecutionFailureInfo`?

Reply:
Do you mean something like this?
I want to choose the method which presto team think best.
```diff
diff --git a/presto-client/src/main/java/io/prestosql/client/FailureInfo.java b/presto-client/src/main/java/io/prestosql/client/FailureInfo.java
index a64b09cedb..ce7ef84e49 100644
--- a/presto-client/src/main/java/io/prestosql/client/FailureInfo.java
+++ b/presto-client/src/main/java/io/prestosql/client/FailureInfo.java
@@ -37,6 +37,7 @@ public class FailureInfo
     private final List<FailureInfo> suppressed;
     private final List<String> stack;
     private final ErrorLocation errorLocation;
+    private final String semanticErrorCodeString;

     @JsonCreator
     public FailureInfo(
@@ -45,7 +46,8 @@ public class FailureInfo
             @JsonProperty("cause") FailureInfo cause,
             @JsonProperty("suppressed") List<FailureInfo> suppressed,
             @JsonProperty("stack") List<String> stack,
-            @JsonProperty("errorLocation") @Nullable ErrorLocation errorLocation)
+            @JsonProperty("errorLocation") @Nullable ErrorLocation errorLocation,
+            @JsonProperty("semanticErrorCodeString") String semanticErrorCodeString)
     {
         requireNonNull(type, "type is null");
         requireNonNull(suppressed, "suppressed is null");
@@ -57,6 +59,7 @@ public class FailureInfo
         this.suppressed = ImmutableList.copyOf(suppressed);
         this.stack = ImmutableList.copyOf(stack);
         this.errorLocation = errorLocation;
+        this.semanticErrorCodeString = semanticErrorCodeString;
     }

     @JsonProperty
@@ -98,6 +101,13 @@ public class FailureInfo
         return errorLocation;
     }

+    @Nullable
+    @JsonProperty
+    public String getSemanticErrorCodeString()
+    {
+        return semanticErrorCodeString;
+    }
+
     public RuntimeException toException()
     {
         return toException(this);
diff --git a/presto-main/src/main/java/io/prestosql/execution/ExecutionFailureInfo.java b/presto-main/src/main/java/io/prestosql/execution/ExecutionFailureInfo.java
index 29e9a240a9..644ed215ab 100644
--- a/presto-main/src/main/java/io/prestosql/execution/ExecutionFailureInfo.java
+++ b/presto-main/src/main/java/io/prestosql/execution/ExecutionFailureInfo.java
@@ -20,6 +20,7 @@ import io.prestosql.client.ErrorLocation;
 import io.prestosql.client.FailureInfo;
 import io.prestosql.spi.ErrorCode;
 import io.prestosql.spi.HostAddress;
+import io.prestosql.sql.analyzer.SemanticErrorCode;

 import javax.annotation.Nullable;
 import javax.annotation.concurrent.Immutable;
@@ -45,6 +46,7 @@ public class ExecutionFailureInfo
     private final ErrorCode errorCode;
     // use for transport errors
     private final HostAddress remoteHost;
+    private final SemanticErrorCode semanticErrorCode;

     @JsonCreator
     public ExecutionFailureInfo(
@@ -55,7 +57,8 @@ public class ExecutionFailureInfo
             @JsonProperty("stack") List<String> stack,
             @JsonProperty("errorLocation") @Nullable ErrorLocation errorLocation,
             @JsonProperty("errorCode") @Nullable ErrorCode errorCode,
-            @JsonProperty("remoteHost") @Nullable HostAddress remoteHost)
+            @JsonProperty("remoteHost") @Nullable HostAddress remoteHost,
+            @JsonProperty("semanticErrorCode") @Nullable SemanticErrorCode semanticErrorCode)
     {
         requireNonNull(type, "type is null");
         requireNonNull(suppressed, "suppressed is null");
@@ -69,6 +72,7 @@ public class ExecutionFailureInfo
         this.errorLocation = errorLocation;
         this.errorCode = errorCode;
         this.remoteHost = remoteHost;
+        this.semanticErrorCode = semanticErrorCode;
     }

     @JsonProperty
@@ -124,13 +128,20 @@ public class ExecutionFailureInfo
         return remoteHost;
     }

+    @Nullable
+    @JsonProperty
+    public SemanticErrorCode getSemanticErrorCode()
+    {
+        return semanticErrorCode;
+    }
+
     public FailureInfo toFailureInfo()
     {
         List<FailureInfo> suppressed = this.suppressed.stream()
                 .map(ExecutionFailureInfo::toFailureInfo)
                 .collect(toImmutableList());

-        return new FailureInfo(type, message, cause == null ? null : cause.toFailureInfo(), suppressed, stack, errorLocation);
+        return new FailureInfo(type, message, cause == null ? null : cause.toFailureInfo(), suppressed, stack, errorLocation, semanticErrorCode.name());
     }

     public RuntimeException toException()
diff --git a/presto-main/src/main/java/io/prestosql/execution/SqlStageExecution.java b/presto-main/src/main/java/io/prestosql/execution/SqlStageExecution.java
index 0953769a5a..baf8c0c3ce 100644
--- a/presto-main/src/main/java/io/prestosql/execution/SqlStageExecution.java
+++ b/presto-main/src/main/java/io/prestosql/execution/SqlStageExecution.java
@@ -542,7 +542,8 @@ public final class SqlStageExecution
                 executionFailureInfo.getStack(),
                 executionFailureInfo.getErrorLocation(),
                 REMOTE_HOST_GONE.toErrorCode(),
-                executionFailureInfo.getRemoteHost());
+                executionFailureInfo.getRemoteHost(),
+                executionFailureInfo.getSemanticErrorCode());
     }

     @Override
diff --git a/presto-main/src/main/java/io/prestosql/util/Failures.java b/presto-main/src/main/java/io/prestosql/util/Failures.java
index 4e6da1d810..cd9120e8ee 100644
--- a/presto-main/src/main/java/io/prestosql/util/Failures.java
+++ b/presto-main/src/main/java/io/prestosql/util/Failures.java
@@ -24,6 +24,7 @@ import io.prestosql.spi.HostAddress;
 import io.prestosql.spi.PrestoException;
 import io.prestosql.spi.PrestoTransportException;
 import io.prestosql.spi.StandardErrorCode;
+import io.prestosql.sql.analyzer.SemanticErrorCode;
 import io.prestosql.sql.analyzer.SemanticException;
 import io.prestosql.sql.parser.ParsingException;
 import io.prestosql.sql.tree.NodeLocation;
@@ -96,7 +97,7 @@ public final class Failures
         }

         if (seenFailures.contains(throwable)) {
-            return new ExecutionFailureInfo(type, "[cyclic] " + throwable.getMessage(), null, ImmutableList.of(), ImmutableList.of(), null, GENERIC_INTERNAL_ERROR.toErrorCode(), remoteHost);
+            return new ExecutionFailureInfo(type, "[cyclic] " + throwable.getMessage(), null, ImmutableList.of(), ImmutableList.of(), null, GENERIC_INTERNAL_ERROR.toErrorCode(), remoteHost, null);
         }
         seenFailures.add(throwable);

@@ -111,6 +112,12 @@ public final class Failures
             }
         }

+        SemanticErrorCode semanticErrorCode = null;
+        if (throwable instanceof SemanticException) {
+            SemanticException e = (SemanticException) throwable;
+            semanticErrorCode = e.getCode();
+        }
+
         return new ExecutionFailureInfo(
                 type,
                 throwable.getMessage(),
@@ -121,7 +128,8 @@ public final class Failures
                 Lists.transform(asList(throwable.getStackTrace()), toStringFunction()),
                 getErrorLocation(throwable),
                 errorCode,
-                remoteHost);
+                remoteHost,
+                semanticErrorCode);
     }

     @Nullable
```

Reply:
Much better.

Reply:
Add this as a new field `semanticErrorType` in `QueryError`.

`FailureInfo` is effectively a generic serialized Java exception (with stack trace and recursive cause and suppressed exceptions).